deepchem/models/normalizing_flows.py

```diff
@@ -0,0 +1,347 @@
+"""
+Normalizing flows for transforming probability distributions.
+"""
+
+import numpy as np
+import logging
+from typing import List, Iterable, Optional, Tuple, Sequence, Any
+
+import tensorflow as tf
+
+import deepchem as dc
+from deepchem.models.losses import Loss
+from deepchem.models.models import Model
+from deepchem.models.optimizers import Adam
+
+logger = logging.getLogger(__name__)
+
+
+class NormalizingFlow(tf.keras.models.Model):
+  """Base class for normalizing flow.
+
+  The purpose of a normalizing flow is to map a simple distribution (that is
+  easy to sample from and evaluate probability densities for) to a more
+  complex distribution that is learned from data. The base distribution 
+  p(x) is transformed by the associated normalizing flow y=g(x) to model the
+  distribution p(y).
+
+  Normalizing flows combine the advantages of autoregressive models
+  (which provide likelihood estimation but do not learn features) and
+  variational autoencoders (which learn feature representations but
+  do not provide marginal likelihoods).
+
+  """
+
+  def __init__(self, **kwargs):
+    """Create a new NormalizingFlow."""
+
+    super(NormalizingFlow, self).__init__(**kwargs)
+
+    # An instance of tfd.TransformedDistribution
+    self.flow = None
+
+  def __call__(self, *x):
+    return self.flow.bijector.forward(*x)
+
+  @tf.function
+  def fit_on_batch(self, x: np.ndarray,
+                   optimizer: tf.keras.optimizers.Optimizer,
+                   loss: dc.models.losses.Loss) -> float:
+    """Fit on batch of samples.
+    
+    Parameters
+    ----------
+    x: np.ndarray, shape (n_samples, n_dim)
+      Array of samples where each sample is a vector of length `n_dim`.
+    optimizer: dc.models.optimizers.Optimizer
+      An instance of Optimizer.
+    loss: dc.models.losses.Loss
+      An instance of Loss.
+
+    Returns
+    -------
+    batch_loss: float
+      Loss computed on this batch.
+
+    """
+
+    with tf.GradientTape() as tape:
+      dummy_labels = np.ones(len(x))
+      batch_loss = loss(x, dummy_labels)
+      grads = tape.gradient(batch_loss, self.trainable_variables)
+      optimizer.apply_gradients(zip(grads, self.trainable_variables))
+    return batch_loss
+
+
+class NormalizingFlowModel(NormalizingFlow):
```

Comment:
Would it be possible for NormalizingFlowModel to be a subclass of `dc.models.Model`?